TombEngine/Objects/TR3/Entity/FishSwarm.cpp

```diff
@@ -0,0 +1,410 @@
+
+		// 1) Test for water room.
+		if (!TestEnvironment(ENV_FLAG_WATER, pointColl.RoomNumber))
+			return Vector3::Zero;
+
+		// 2) Assess point collision.
+		if (pos.y >= (pointColl.Position.Floor - BUFFER) ||
+			pos.y <= (waterHeight + BUFFER) ||
+			pointColl.Block->IsWall(item.Pose.Position.x + BUFFER, item.Pose.Position.z + BUFFER) ||
+			pointColl.Block->IsWall(item.Pose.Position.x - BUFFER, item.Pose.Position.z - BUFFER))
+		{
+			return Vector3::Zero;
+		}
+
+		return pos;
+	}
+
+	void UpdateFishSwarm()
+	{
+		constexpr auto WATER_SURFACE_OFFSET = CLICK(0.5f);
+		constexpr auto FLEE_VEL				= 20.0f;
+
+		static const auto SPHERE = BoundingSphere(Vector3::Zero, BLOCK(1 / 8.0f));
+
+		if (FishSwarm.empty())
+			return;
+
+		const FishData* closestFishPtr = nullptr;
+		float minDistToTarget = INFINITY;
+		int minDist = INT_MAX;
+
+		int fishNumber = 0;
+		for (auto& fish : FishSwarm)
+		{
+			if (fish.Life <= 0.0f)
+				continue;
+
+			// Increase separation distance for each fish.
+			float separationDist = FISH_BASE_SEPARATION_DISTANCE + (fishNumber * 3);
+			fishNumber += 1;
+
+			auto& leaderItem = *fish.LeaderItemPtr;
+
+			if (!leaderItem.ItemFlags[2] && fish.TargetItemPtr == fish.LeaderItemPtr)
+			{
+				if (!fish.IsPatrolling)
+				{
+					fish.TargetItemPtr->Pose.Position = GetFishStartPosition(leaderItem);
+
+					if (fish.TargetItemPtr->Pose.Position != Vector3::Zero)
+						leaderItem.ItemFlags[2] = 1;
+				}
+			}
+
+			int enemyVel = (fish.TargetItemPtr != fish.LeaderItemPtr) ? 16 : 26;
+
+			fish.PositionTarget = Random::GeneratePointInSphere(SPHERE);
+
+			// Calculate desired position based on target object and random offsets.
+			auto desiredPos = fish.TargetItemPtr->Pose.Position + fish.PositionTarget;
+			auto dir = desiredPos - fish.Position;
+
+			auto dirs = dir.ToVector3();
+			dirs.Normalize();
+			auto dirNorm = dirs;
+
+			// Define cohesion factor to keep fish close together.
+			float distToTarget = dirs.Length();
+
+			float targetVel = (distToTarget * FISH_COHESION_FACTOR) + Random::GenerateFloat(3.0f, 5.0f);
+			fish.Velocity = std::min(targetVel, fish.TargetItemPtr->Animation.Velocity.z - 21.0f); 
+
+			// If fish is too far from target, increase velocity to catch up.
+			if (distToTarget > FISH_TARGET_DISTANCE_MAX)
+				fish.Velocity += FISH_CATCH_UP_FACTOR; 
+
+			// Translate.
+			auto moveDir = fish.Orientation.ToDirection();
+			moveDir.Normalize(); 
+			fish.Position += (moveDir * fish.Velocity / enemyVel);
+			fish.Position += (moveDir * FISH_SPACING_FACTOR) / enemyVel;
+
+			auto orientTo = Geometry::GetOrientToPoint(fish.Position, desiredPos.ToVector3());
+			fish.Orientation.Lerp(orientTo, 0.1f);
+
+			for (const auto& otherFish : FishSwarm)
+			{
+				if (&fish == &otherFish)
+					continue;
+
+				float distToOtherFish = Vector3i::Distance(fish.Position, otherFish.Position);
+				float distToPlayer = Vector3i::Distance(fish.Position, LaraItem->Pose.Position);
```

Comment:
Probably a note for the future, but if we somehow manage to detect corpses to attack by fish swarm, maybe we can also detect any creature in the water, not only Lara? Then you can arrange different scenarios, such as shark or crocodile attacked by pirahnas, and so on.